src/test/regress/sql/with.sql

```diff
@@ -1083,3 +1083,41 @@ WITH cte AS (
 
 RESET optimizer;
 DROP TABLE d;
+
+-- Check if sharing is disabled for a SegmentGeneral CTE to avoid deadlock if CTE is
+-- executed with 1-gang and joined with Hashed
+SET optimizer = off;
+--start_ignore
+DROP TABLE IF EXISTS d;
+DROP TABLE IF EXISTS r;
+--end_ignore
+
+CREATE TABLE d (a int, b int) DISTRIBUTED BY (a);
+INSERT INTO d VALUES (1,2),(2,3);
+CREATE TABLE r (a int, b int) DISTRIBUTED REPLICATED;
+INSERT INTO r VALUES (1,2),(3,4);
+
+EXPLAIN (COSTS off)
+WITH cte AS (
+    SELECT count(*) a FROM r
+) SELECT * FROM cte JOIN (SELECT * FROM d JOIN cte USING (a) LIMIT 1) d_join_cte USING (a);
+
+WITH cte AS (
+    SELECT count(*) a FROM r
+) SELECT * FROM cte JOIN (SELECT * FROM d JOIN cte USING (a) LIMIT 1) d_join_cte USING (a);
+
+-- Check if sharing is still enabled for other cases
+ALTER TABLE r SET DISTRIBUTED BY (a);
```

Comment:
Instead, you can simply replace `r` with `d` in the query to make two `d`'s. And another question: what about a query with two `r`?

Reply:
The correct plan generates:
```
                         QUERY PLAN                          
-------------------------------------------------------------
 Hash Join
   Hash Cond: (d_join_cte.a = (count(*)))
   ->  Subquery Scan on d_join_cte
         ->  Limit
               ->  Gather Motion 1:1  (slice1; segments: 1)
                     ->  Hash Join
                           Hash Cond: (r.a = (count(*)))
                           ->  Seq Scan on r
                           ->  Hash
                                 ->  Aggregate
                                       ->  Seq Scan on r r_1
   ->  Hash
         ->  Gather Motion 1:1  (slice2; segments: 1)
               ->  Aggregate
                     ->  Seq Scan on r r_2
```
This is the case where Shared Scan can be used, however, as stated before, we don't know what the shared scan will be joined with when building CTE plan.

Should add this case to the test instead of joining two Hashed.